### DISCUSSION TITLE
Change the sender of a Lockup stream

### DISCUSSION THREAD

**Discussion by PaulRBerg** [sablier-labs/evm-monorepo#1030] [Ideas]
Requested by a user recently:

<img width="350" alt="SCR-20240902-lqdg" src="https://github.com/user-attachments/assets/d7b9c90a-bb1a-4e42-84a6-eca096658b27">

IIRC:

1. We have received episodic requests for this feature over the years. Right, @maxdesalle?
2. Hedgey offers this feature already.

RFC @sablier-labs/everybody

**Comment by maxdesalle** [comment]
Correct! This could be under the form of a new sender NFT, that would carry the right to cancel a stream (assuming it's cancelable).

**Comment by smol-ninja** [comment]
Sounds good. Just one point to add, instead of a simple `transferAdmin` function, we can consider implementing it as a two-step transfer to mitigate the risk of accidental transfer to an incorrectly copied address. It was also mentioned in our Cantina audit ([3.3.1](https://github.com/sablier-labs/audits/blob/main/v2-core/cantina-2024-06-13.pdf)).

**Reply by PaulRBerg** [reply]
I think you're referring to something else?

We're talking about stream senders here.

**Reply by smol-ninja** [reply]
I am talking about the stream sender. IIUC, the requirement is to allow a stream sender to update the stream's sender address, right?

**Reply by PaulRBerg** [reply]
Yes. But above, you said `transferAdmin`. I thought you thought this was about transferring the admin.

Anyway. I don't think a two-step process would be helpful here. Like, transferring the recipient is a one-step process, which is arguably more dangerous than transferring the sender (i.e. the ability to cancel).

**Reply by smol-ninja** [reply]
> transferring the recipient is a one-step process, which is arguably more dangerous than transferring the sender

Yeah good point.

**Reply by andreivladbrg** [reply]
> transferring the sender (i.e. the ability to cancel).

that's a matter of what's the elapsed time in the stream. if the stream just started and it is cancelable, basically, the sender still owns more of that stream.

re. the proposal, i think we can do it. there isn't anything atm that comes to mind that would cause security issues.

**Reply by PaulRBerg** [reply]
> that's a matter of what's the elapsed time in the stream

Yeah, good point that it depends.

**Comment by maxdesalle** [comment]
Was just on an unscheduled call with the Fjord team as they have a compromised sender wallet, and are trying to see how they could mass-renounce the streams. Could have been much easier if they could have transferred ownership on the sender side of the streams to another wallet. Batch stream transfers would require Account Abstraction, though, so this is probably not immediately urgent.

**Reply by PaulRBerg** [reply]
As discussed on Slack, a sender transfer feature would have been helpful only if they could batch transfer streams.

Also, in the case where the attacker had been aware of the streams, the transfer feature wouldn't have achieved much.

A more immediate problem is the lack of `renounceMultiple`. Will open a separate discussion about this.

**Comment by PaulRBerg** [comment]
One more (implicit) request for this feature.

<img width="350" alt="SCR-20241007-lrni" src="https://github.com/user-attachments/assets/1f8d2c2c-2b46-424e-9942-dd533e21e072">

**Comment by PaulRBerg** [comment]
Request from Ondefy:

<img width="400" alt="SCR-20241127-pntz" src="https://github.com/user-attachments/assets/bbc816c2-ff99-4d3b-aac3-334680743128">

**Comment by smol-ninja** [comment]
Request from YieldNest:

<img width="568" alt="Screenshot 2025-06-02 at 12 36 23" src="https://github.com/user-attachments/assets/4e4a09aa-11ee-483b-bcbc-b90bcc3dd506" />
